base/base/itoa.cpp

```diff
@@ -137,161 +24,260 @@ const char digits[201] = "00010203040506070809"
+    pair fd[100]{
+        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', //
+        10,  11,  12,  13,  14,  15,  16,  17,  18,  19, //
+        20,  21,  22,  23,  24,  25,  26,  27,  28,  29, //
+        30,  31,  32,  33,  34,  35,  36,  37,  38,  39, //
+        40,  41,  42,  43,  44,  45,  46,  47,  48,  49, //
+        50,  51,  52,  53,  54,  55,  56,  57,  58,  59, //
+        60,  61,  62,  63,  64,  65,  66,  67,  68,  69, //
+        70,  71,  72,  73,  74,  75,  76,  77,  78,  79, //
+        80,  81,  82,  83,  84,  85,  86,  87,  88,  89, //
+        90,  91,  92,  93,  94,  95,  96,  97,  98,  99, //
+    };
+} digits;
+
+constexpr UInt64 mask24 = (UInt64(1) << 24) - 1;
+constexpr UInt64 mask32 = (UInt64(1) << 32) - 1;
+constexpr UInt64 mask57 = (UInt64(1) << 57) - 1;
+
+template <bool, class, class F>
+struct _cond
 {
-    return outTwoDigits(p, u);
-}
-
-//===----------------------------------------------------------===//
-// large values are >= 10^2*N
-// where x contains quotient and remainder after division by 10^N
-//===----------------------------------------------------------===//
-template <size_t N>
-ALWAYS_INLINE inline char * large(char * p, QuotientAndRemainder<N> x)
+    using type = F;
+};
+template <class T, class F>
+struct _cond<true, T, F>
 {
-    QuotientAndRemainder<N> y = split<N>(x.quotient);
-    p = head(p, UnsignedOfSize<N / 2>(y.quotient));
-    p = tail(p, y.remainder);
-    p = tail(p, x.remainder);
-    return p;
-}
+    using type = T;
+};
+template <bool B, class T, class F>
+using cond = typename _cond<B, T, F>::type;
 
-//===----------------------------------------------------------===//
-// handle values of "u" that might be >= 10^2*N
-// where N is the size of "u" in bytes
-//===----------------------------------------------------------===//
-template <typename UInt, size_t N = sizeof(UInt)>
-ALWAYS_INLINE inline char * uitoa(char * p, UInt u)
+template <class T>
+inline ALWAYS_INLINE char * to_text_from_integer(char * b, T i)
 {
-    if (u < pow10<UnsignedOfSize<N>>(N))
-        return head(p, UnsignedOfSize<N / 2>(u));
-    QuotientAndRemainder<N> x = split<N>(u);
+    constexpr auto q = sizeof(T);
+    using U = cond<q == 1, char8_t, cond<q <= sizeof(UInt16), UInt16, cond<q <= sizeof(UInt32), UInt32, UInt64>>>;
 
-    return u < pow10<UnsignedOfSize<N>>(2 * N) ? head<N>(p, x) : large<N>(p, x);
-}
+    // convert bool to int before test with unary + to silence warning if T happens to be bool
+    U const n = +i < 0 ? *b++ = '-', U(0) - U(i) : U(i);
 
-// selected when "u" is one byte
-template <>
-ALWAYS_INLINE inline char * uitoa<UnsignedOfSize<1>, 1>(char * p, UnsignedOfSize<1> u)
-{
-    if (u < 10)
-        return outDigit(p, u);
-    else if (u < 100)
-        return outTwoDigits(p, u);
-    else
+    if (n < U(1e2))
     {
-        p = outDigit(p, u / 100);
-        p = outTwoDigits(p, u % 100);
-        return p;
+        return n < 10 ? outOneDigit(b, n) : outTwoDigits(b, n);
```

Comment:
Shouldn't we add a comment that it's our modification compared to the original implementation?
Also, it's interesting why it is faster than `*reinterpret_cast<pair *>(b) = digits.fd[n]`.

Reply:
> Also, it's interesting why it is faster than *reinterpret_cast<pair *>(b) = digits.fd[n].

It's not simple to say because it's not something you would expect based on the code. Here is an extract when printing small numbers (<10), which is where this library was slower than what we had before:

Before:
```
       │     Disassembly of section .text:                                                                                                                                                                                                                                              
       │                                                                                                                                                                                                                                                                                
       │     0000000014fd6620 <itoa(unsigned long, char*)>:                                                                                                                                                                                                                             
 12.43 │       cmp    $0x63,%rdi                                                                                                                                                                                                                                                        
       │     ↓ ja     21                                                                                                                                                                                                                                                                
 10.93 │       movzwl 0x2530240(%rdi,%rdi,1),%eax                                                                                                                                                                                                                                       
 11.66 │       mov    %ax,(%rsi)                                                                                                                                                                                                                                                        
  8.14 │       cmp    $0xa,%rdi                                                                                                                                                                                                                                                         
 11.50 │       sbb    $0x0,%rsi                                                                                                                                                                                                                                                         
 11.99 │       add    $0x2,%rsi                                                                                                                                                                                                                                                         
  9.65 │       mov    %rsi,%rax                                                                                                                                                                                                                                                         
 23.70 │     ← ret                                                                                                                                                                                                                                                                      
       │ 21:   cmp    $0xf423f,%rdi 
...
```

After:
```
       │     Disassembly of section .text:                                                                                                                                                                                                                                              
       │                                                                                                                                                                                                                                                                                
       │     0000000014fd66c0 <itoa(unsigned long, char*)>:                                                                                                                                                                                                                             
  4.44 │       cmp    $0x63,%rdi                                                                                                                                                                                                                                                        
       │     ↓ ja     1a                                                                                                                                                                                                                                                                
 13.57 │       cmp    $0x9,%rdi                                                                                                                                                                                                                                                         
       │     ↓ ja     79                                                                                                                                                                                                                                                                
 18.36 │       add    $0x30,%dil                                                                                                                                                                                                                                                        
 18.51 │       mov    %dil,(%rsi)                                                                                                                                                                                                                                                       
  9.05 │       inc    %rsi                                                                                                                                                                                                                                                              
       │       mov    %rsi,%rax                                                                                                                                                                                                                                                         
 36.08 │     ← ret                                                                                                                                                                                                                                                                      
       │ 1a:   nop                                                                                                                                                                                                                                                                      
...
```

To print small numbers the upstream code did 1 branch + 6 instructions (one of them requires access to a table). The new code does 2 branches + 4 instructions (no memory access required). If the branch predictor is good then the extra branch cost less than what we save by removing 2 instructions and one memory access.
Mostly this works because numbers in a real life dataset aren't random.

I'll add a comment